Keep case search query inputs out of address bar

nospame opened the pull request:
## Product Description
Case search "inputs" are no longer in the web address bar, so if any projects copy + paste links directly to case search results this will no longer be possible.

## Technical Summary
[USH-3834](https://dimagi-dev.atlassian.net/browse/USH-3834)
Rather than storing search inputs in the url, this change means that query inputs are now stored in and retrieved from `Utils.queryInputs`, under their appropriate `queryKey` to account for times when more than one set of queryData may be present. The same data was previously already stored as `stickyQueryInputs`, so there is no added security concern. This change keeps `stickyQueryInputs` functionality with minor refactoring, as it represents a different state of inputs -- one that is used to repopulate previously entered values, rather than used to construct a request to Formplayer.

## Feature Flag
SYNC_SEARCH_CASE_CLAIM

## Safety Assurance

### Safety story
Tested locally with different case search workflows (split screen & regular) and found basic case search works as expected and keeps query inputs out of the browser URL. Tested on staging with various parent > child select and multi-select workflows, which surfaced issues with the original approach that have now been addressed.

### Automated test coverage
Touched by mocha test "should navigate through case search" showing that the navigation url is ultimately set correctly.

### QA Plan
This should get QA to ensure compatibility with various case search workflows and split screen case search with or without dynamic search.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[USH-3834]: https://dimagi-dev.atlassian.net/browse/USH-3834?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
Keep case search query inputs out of address bar

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/router.js:
```diff
@@ -201,18 +201,21 @@ hqDefine("cloudcare/js/formplayer/router", function () {
+            queryObject.forceManualSearch = true;
+        }
         urlObject.setQueryData(queryObject);
         utils.setUrlToObject(urlObject);
-        API.listMenus();
+
+        // add inputs after setting browser url to keep sensitive info out of address bar
+        queryObject.inputs = queryDict;
+        urlObject.setQueryData(queryObject);
```
I agree that it looks kind of wonky but I don't think there is a more concise way.

Jtang-1 replied:
thought about it for a while and couldn't come up with a better way either

MartinRiese commented:
@nospame I had I PR I could not merge too. I could not figure out why it was disabling the merge button. I ended up opening a new PR with the same changes.

nospame commented:
Having to rethink this one somewhat since it's causing issues with some other web apps navigation that relies on `currentUrlToObject` including `inputs`

nospame committed:
Revert "Keep case search query inputs out of address bar"

This reverts commit a6db3791773ab46af4ffb2b32d04cbc703822933.

nospame committed:
Store case search query inputs in sessionStorage, not url

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into em/obfuscate-case-search-inputs

nospame committed:
Add inputs back to queryData

nospame committed:
Check if queryData[queryKey] exists before writing to it

nospame committed:
Refactor: get/set query inputs with separate util functions

Also clear queryInputs from session in more places

nospame committed:
Cleanup: remove redundant conditional and comment

nospame committed:
Remove isEmpty check when adding inputs to cloudcareUrl

In cases such as clearing previous search terms, empty inputs are intentional

nospame committed:
Don't retrieve queryInputs for bad queryKey

Jtang-1 reviewed (commented):
The logic to store the inputs keyed by the `queryKey` seems similar to what `stickyQueryInputs` does. Except you're storing it in a sessionStorage and stickyQueryInput seems to just store it in an object instance? I wonder if the functionalities can by combined?

MartinRiese commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -52,15 +52,28 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
-            return Utils.CloudcareUrl.fromJson(Utils.encodedUrlToObject(url));
+            const cloudcareUrl = Utils.CloudcareUrl.fromJson(Utils.encodedUrlToObject(url));
+            const queryKey = sessionStorage.queryKey;
+            if (queryKey && queryKey !== "null" && queryKey !== "undefined") {
```
ooo. I like that. Making a mental note.

nospame replied:
I don't think that's applicable here unless I'm misunderstanding the proposed use. `sessionStorage.queryKey` can potentially be any of: `undefined` (not set), the strings `"null"` or `"undefined"` depending on when it gets checked, or a genuine query key e.g. `search_results.m0`. Only in the lattermost case do we actually want to use its value to set inputs on the `CloudcareUrl` object.

Jtang-1 replied:
I was thinking like 
```
if (queryKey ?? false) {
}
```
But that wouldn't work if `queryKey` can be the strings "null" and "undefined". Is that what's returned if sessionStorage doesn't have a `queryKey`? or do we store those strings for some reason?

nospame commented:
> The logic to store the inputs keyed by the queryKey seems similar to what stickyQueryInputs does. Except you're storing it in a sessionStorage and stickyQueryInput seems to just store it in an object instance? I wonder if the functionalities can by combined?
@Jtang-1 That's a good point, it's possible this doesn't need to use `sessionStorage` at all if the Utils object is persistent and accessible. I'll take a look at that.

nospame committed:
Store query inputs on Utils object

No need for sessionStorage

nospame committed:
Minimize calls to sessionStorage for queryKey

nospame committed:
Absorb setStickyQueryInputs into setQueryInputs

nospame committed:
Clear queryInputs when queryData cleared

nospame committed:
Don't store nullish-ish keys on queryInputs

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -185,29 +194,31 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
+        queryKey = queryKey || sessionStorage.queryKey;
+        if (queryKey && queryKey !== "null" && queryKey !== "undefined") {
+            queryInputs[queryKey] = inputs;
+            this.queryInputs = queryInputs;
```
nit: can this just be `this.queryInputs[queryKey] = inputs`

nospame replied:
Good call, though I do at least have to check that `this.queryInputs` isn't undefined to set a key on it. What would you think about
```javascript
            this.queryInputs = this.queryInputs || {};
            this.queryInputs[queryKey] = inputs;
```
I could also call `getQueryInputs()` which is more DRY, but I think it's more efficient just to assign it to itself (or `{}` ) than assigning the result of a function.

Jtang-1 replied:
ah good catch for checking if `this.queryInputs` is undefined. Change look good

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -324,6 +337,7 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
             this.search = null;
             this.queryData = null;
             this.sessionId = null;
+            Utils.clearQueryInputs();
```
Should we make this a `clearQueryData` and throw `this.queryData` in there as well?

nospame replied:
I realized that clearing the query inputs actually breaks sticky query inputs. That data was being stored whether or not the feature flag was enabled -- the flag just controls whether it's pulled from or not. So I reverted that commit. https://github.com/dimagi/commcare-hq/pull/33946/commits/5669124cef39ec39045407913fb5d93bbd59bb34

Jtang-1 replied:
that makes sense

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -52,7 +52,12 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
+            const cloudcareUrl = Utils.CloudcareUrl.fromJson(Utils.encodedUrlToObject(url));
+            for (const queryKey in cloudcareUrl.queryData) {
+                // retrieve query inputs from Utils object
+                cloudcareUrl.queryData[queryKey].inputs = Utils.getCurrentQueryInputs(queryKey);
```
do we also need to set queryInputs here?

nospame replied:
I think "no" since it's already being set here https://github.com/dimagi/commcare-hq/blob/5669124cef39ec39045407913fb5d93bbd59bb34/corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js#L299 whenever we intentionally `setQueryData` on the cloudcareUrl object

Jtang-1 replied:
I'm looking at how currentUrlToObject is used [here](https://github.com/dimagi/commcare-hq/blob/a7a0f81ca4e2f026f0f0a1c0d0cd86c656ced6a6/corehq/apps/cloudcare/static/cloudcare/js/formplayer/app.js#L438-L444) and it accesses `.queryData`. In this case, the queryData.input would be missing right?

nospame replied:
I believe `queryData.inputs` should only be missing if it was never set in the first place, and in those cases we wouldn't expect to find it anyway.

Jtang-1 replied:
Ah I think you're right, I was getting this confused with something else. This looks good

Jtang-1 reviewed (commented):
nice! looking good, just a couple comments

nospame committed:
Revert "Clear queryInputs when queryData cleared"

Otherwise sticky query inputs would not be sticky. This reverts commit 49d4f48bc4c79f6b076892c5bfd87d6e2bb2e573.

nospame committed:
Refactor: simplify setQueryInputs assignment

nospame committed:
Don't return undefined from getCurrentQueryInputs

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into em/obfuscate-case-search-inputs

nospame committed:
Only set query inputs with query data if inputs not empty

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -276,20 +282,23 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
             }
-
-            this.queryData[sessionStorage.queryKey] = queryDataEntry;
+            if (!_.isEmpty(inputs)) {
```
should this be handled inside `setQueryInputs` and set value of inputs to `{}` if it's empty?

Otherwise, would this check also need to be done in `query.js` for 
`formplayerUtils.setQueryInputs(this.getAnswers());`

nospame replied:
I don't think so, but I also don't think the check I have is quite right. Sometimes `inputs` needs to be intentionally set to `null` or an empty object `{}`. I was hoping not to have to investigate every place this is called, but I think I do need to in working on an issue that's occurring now between default values and sticky query inputs.

nospame committed:
Split stickyQueryInputs back out from queryInputs

They're very similar but represent two different things

nospame committed:
Remove null checks for queryInputs

null is sometimes needed, e.g. in setting inputs null to retrieve default values from Formplayer

nospame committed:
Refactor: rename, reorder for consistency

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -198,36 +194,33 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
-        return queryInputs[sessionStorage.queryKey] || {};
+    Utils.setCurrentQueryInputs = function (inputs, queryKey) {
+        queryKey = queryKey || sessionStorage.queryKey;
+        if (queryKey && queryKey !== "null" && queryKey !== "undefined") {
```
Bit of discussion with @Jtang-1 on this above: https://github.com/dimagi/commcare-hq/pull/33946#discussion_r1449366455

Basically, without changing the logic of how/when `sessionStorage.queryKey` gets set which seems outside the scope of this PR, it sometimes gets set from a JS `undefined` or `null` to `"undefined"` or `"null"` since sessionStorage only holds strings. I could remove the check, it's not technically necessary, but it seemed nicer to avoid storing values we can't accurately look up later.

nospame commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/utils/utils.js:
```diff
@@ -198,36 +194,33 @@ hqDefine("cloudcare/js/formplayer/utils/utils", function () {
     Utils.setStickyQueryInputs = function (inputs) {
-        if (!this.stickyQueryInputs) {
-            this.stickyQueryInputs = {};
+        const queryKey = sessionStorage.queryKey;
```
It really does. I tried to refactor `queryInputs` and `stickyQueryInputs` to use the same object but that caused all sorts of weird issues in QA so I split them back out.
